src/main/java/org/jvnet/hudson/test/InboundAgentRule.java

```diff
@@ -321,4 +337,127 @@ public boolean isAlive(String name) {
+        private final File agentJar;
+        /**
+         * The secret the agent should use to connect.
+         */
+        @NonNull
+        private final String secret;
+        /**
+         * The number of nodes in the Jenkins instance where the agent is running.
+         */
+        private final int numberOfNodes;
+        /**
+         * Additional command line arguments to pass to the agent.
+         */
+        @NonNull
+        private final List<String> commandLineArgs;
+
+        public AgentArguments(@NonNull String agentJnlpUrl, @NonNull File agentJar, @NonNull String secret, int numberOfNodes, @NonNull List<String> commandLineArgs) {
+            this.agentJnlpUrl = agentJnlpUrl;
+            this.agentJar = agentJar;
+            this.secret = secret;
+            this.numberOfNodes = numberOfNodes;
+            this.commandLineArgs = commandLineArgs;
+        }
+    }
+
+    private static class GetAgentArguments implements RealJenkinsRule.Step2<AgentArguments> {
+
+        private final String name;
+
+        GetAgentArguments(String name) {
+            this.name = name;
+        }
+
+        @Override
+        public AgentArguments run(@NonNull JenkinsRule r) throws Throwable {
+            return get(r, name);
+        }
+        private static AgentArguments get(JenkinsRule r, String name) throws IOException {
+            Node node = r.jenkins.getNode(name);
+            if (node == null) {
+                throw new AssertionError("no such agent: " + name);
+            }
+            SlaveComputer c = (SlaveComputer) node.toComputer();
+            if (c == null) {
+                throw new AssertionError("agent " + node + " has no executor");
+            }
+            JNLPLauncher launcher = (JNLPLauncher) c.getLauncher();
+            List<String> commandLineArgs = Collections.emptyList();
+            if (!launcher.getWorkDirSettings().isDisabled()) {
+                commandLineArgs = launcher.getWorkDirSettings().toCommandLineArgs(c);
+            }
+            File agentJar = new File(r.jenkins.getRootDir(), "agent.jar");
+            if (!agentJar.isFile()) {
+                FileUtils.copyURLToFile(new Slave.JnlpJar("agent.jar").getURL(), agentJar);
+            }
+            return new AgentArguments(r.getURL() + "computer/" + name + "/slave-agent.jnlp", agentJar, c.getJnlpMac(), r.jenkins.getNodes().size(), commandLineArgs);
+        }
+
+    }
+
+    private static class StopAgent implements RealJenkinsRule.Step {
+        private final String name;
+        StopAgent(String name) {
+            this.name = name;
+        }
+
+        @Override
+        public void run(JenkinsRule r) throws Throwable {
+            stop(r, name);
+        }
+
+        private static void stop(JenkinsRule r, String name) throws InterruptedException {
+            Computer c = r.jenkins.getComputer(name);
+            if (c != null) {
+                while (c.isOnline()) {
+                    Thread.sleep(100);
```

Comment:
The naming implies that this stops the agent, but really it just waits for it go offline?